staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -185,6 +205,25 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	switch requestInfo.Verb {
+	case "create", "update", "patch":
+		found := false
+		for _, v := range crd.Status.StoredVersions {
+			if crdInfo.storageVersion == v {
+				found = true
+				break
+			}
+		}
+		if !found {
+			crd.Status.StoredVersions = append(crd.Status.StoredVersions, crdInfo.storageVersion)
+			crd, err = r.crdGetter.CustomResourceDefinitions().UpdateStatus(crd)
```

Comment:
I don't like this here in the hot path. Why not add a version to the StoredVersions from the create+update strategies in the registry whenever a version is set to "Storage=true"?

Reply:
The more I think about this here, I really dislike it. If we want this to be triggered by the handler, we would need to add a controller with a queue.

Reply:
Actually, moving it into a controller with a queue is not a good idea either: it must be synchronous, i.e. either here or in the registry. Not sure what is actually better.

If we keep it here, please change crdGetter => client.

Reply:
I got the feeling that you were less sure about the first comment's suggestion in later comments. I had the impression that old clients can write in the old version while we are updating the storage but as the CRD specified only one storage version, it is safe to do it in the CRD create/update in the registration. Do you still think it is a good idea? I can move it there.

Reply:
True. With the registry approach we will run into the same problem @xmudrii is trying to solve in https://github.com/kubernetes/kubernetes/pull/63068, namely synchronization of the CRD update and the handlers seeing that change. So I think I can live with this here.

What I want to see:
- a retry loop for conflict errors (and re-check whether the storage version fields is now up-to-date
- a mutex that makes sure that only one handler tries to update the the storage version per process (to avoid multiple handlers doing parallel updates of the same thing).

Reply:
I added the retry but not sure if the mutex would be useful. First it should be a mutex per CRD as we should not block other handlers to update other CRDs then the mutex will only be useful if two handler tries to update the same CRD at the same time. This should be rare and also the storage (etcd) will handle the conflict (i.e. one of the handers will succeed updating the storedVersion, the other one will get a conflict, reads the CRD and realize that the storage version is up to date).